### PR TITLE
Subscriptions API functionality

### PR DESCRIPTION
**Pull request by Mackey22**
* Full endpoint coverage for Subscriptions api
* Full testing of functionality
* Upload files for sample index and query docs for testing
* Manage whitelisted emails with contextmanager and config class

### CODE REVIEW

File: dss-api.yml [hunk 1]
```diff
@@ -687,10 +687,10 @@ definitions:
         description: Replica this query is monitoring. One of aws, gcp, or azure.
         type: string
         enum: [aws, gcp, azure]
-      creator_uid:
-        type: integer
-        format: int
-        description: User ID who created this subscription.
+      owner:
```
Comment (ttung) [inline dss-api.yml]: wait this isn't coming in via oauth?

File: dss/api/subscriptions.py [hunk 2]
```diff
@@ -3,10 +3,14 @@
 import json
 import re
 import typing
+import uuid as uuid_
```
Comment (ttung) [inline dss/api/subscriptions.py]: suggestion:

```
from uuid import uuid4
```

then just use uuid4()

File: dss/api/subscriptions.py [hunk 3]
```diff
@@ -22,16 +26,151 @@
...
 def get(uuid: str, replica: str):
-    pass
+    owner = request.token_info['email']
+
+    es_client = ElasticsearchClient.get(logger)
+    id_exists = es_client.exists(index=DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME,
+                                 doc_type=DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE,
+                                 id=uuid)
+    if not id_exists:
+        return jsonify(dict(message="Subscription {} does not exist".format(uuid),
```
Comment (ttung) [inline dss/api/subscriptions.py]: `raise DSSException(..)`

File: dss/api/subscriptions.py [hunk 4]
```diff
@@ -22,16 +26,151 @@
 
 
 def get(uuid: str, replica: str):
```
Comment (ttung) [inline dss/api/subscriptions.py]: `@dss_handler`

File: dss/api/subscriptions.py [hunk 5]
```diff
@@ -22,16 +26,151 @@
...
+    if not id_exists:
+        return jsonify(dict(message="Subscription {} does not exist".format(uuid),
+                            exception="Placeholder",
+                            HTTPStatusCode=requests.codes.not_found)), requests.codes.not_found
+
+    response = es_client.get(index=DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME,
+                             doc_type=DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE,
+                             id=uuid)
+
+    source = response['_source']
```
Comment (ttung) [inline dss/api/subscriptions.py]: i didn't read the spec, but the key really starts with an underscore?

File: dss/api/subscriptions.py [hunk 6]
```diff
@@ -22,16 +26,151 @@
...
+    source['uuid'] = uuid
+    source['replica'] = replica
+
+    if source['owner'] != owner:
+        return jsonify(dict(
```
Comment (ttung) [inline dss/api/subscriptions.py]: `DSSException`

File: dss/api/subscriptions.py [hunk 7]
```diff
@@ -22,16 +26,151 @@
...
+            message="You don't own subscription {}.".format(uuid),
+            exception="Placeholder",
+            HTTPStatusCode=requests.codes.forbidden)), requests.codes.forbidden
+
+    return jsonify(source), requests.codes.okay
```
Comment (ttung) [inline dss/api/subscriptions.py]: I don't think you need the `requests.code.okay` part, but I could be wrong.

File: dss/api/subscriptions.py (hunk 3 above)
Comment (ttung) [inline dss/api/subscriptions.py]: use f-style strings.

File: dss/api/subscriptions.py [hunk 8]
```diff
@@ -22,16 +26,151 @@
...
+                             doc_type=DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE,
+                             id=uuid)
+
+    source = response['_source']
+    source['uuid'] = uuid
+    source['replica'] = replica
+
+    if source['owner'] != owner:
+        return jsonify(dict(
+            message="You don't own subscription {}.".format(uuid),
```
Comment (ttung) [inline dss/api/subscriptions.py]: f-style strings

File: dss/api/subscriptions.py [hunk 9]
```diff
@@ -22,16 +26,151 @@
...
-    pass
+    owner = request.token_info['email']
+    es_client = ElasticsearchClient.get(logger)
+
+    search_obj = Search(using=es_client,
+                        index=DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME,
+                        doc_type=DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE)
+    search = search_obj.query({'match': {'owner': owner}})
+
+    responses = []
```
Comment (ttung) [inline dss/api/subscriptions.py]:
```
responses = [{
    'uuid': hit.meta.id,
    'replica': replica,
    'owner': owner,
    'callback_url': hit.callback_url,
    'query': hit.query.to_dict()}
  for hit in search.scan()]
```

File: dss/api/subscriptions.py [hunk 10]
```diff
@@ -22,16 +26,151 @@
...
+    }
+    get_elasticsearch_index(es_client, DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME, logger, index_mapping)
+
+    percolate_registration = _register_percolate(es_client, uuid, query)
+
+    if percolate_registration['created']:
+        logger.debug("Percolate query registration succeeded:\n{}".format(percolate_registration))
+    else:
+        logger.critical("Percolate query registration failed:\n{}".format(percolate_registration))
+        return (jsonify(dict(
```
Comment (ttung) [inline dss/api/subscriptions.py]: `DSSException`

File: dss/api/subscriptions.py [hunk 11]
```diff
@@ -22,16 +26,151 @@
...
+            exception="Placeholder",
+            HTTPStatusCode=requests.codes.internal_server_error)),
+            requests.codes.internal_server_error)
+
+    extras['owner'] = owner
+    subscription_registration = _register_subscription(es_client, uuid, extras)
+    if subscription_registration['created']:
+        logger.debug("Event Subscription succeeded:\n{}".format(subscription_registration))
+    else:
+        logger.critical("Event Subscription failed:\n{}".format(subscription_registration))
```
Comment (ttung) [inline dss/api/subscriptions.py]: in the case of failure, is it desirable to remove the percolate query?

File: dss/api/subscriptions.py [hunk 12]
```diff
@@ -22,16 +26,151 @@
...
+        return (jsonify(dict(
```
Comment (ttung) [inline dss/api/subscriptions.py]: DSSException

File: dss/api/subscriptions.py [hunk 13]
```diff
@@ -22,16 +26,151 @@
...
+    if source['owner'] != owner:
+        return jsonify(dict(
+            message="You don't own subscription {}.".format(uuid),
+            exception="Placeholder",
+            HTTPStatusCode=requests.codes.forbidden)), requests.codes.forbidden
+
+    return jsonify(source), requests.codes.okay
 
 
 def find(replica: str):
```
Comment (ttung) [inline dss/api/subscriptions.py]: `@dss_handler`

File: dss/api/subscriptions.py [hunk 14]
```diff
@@ -22,16 +26,151 @@
...
+        return (jsonify(dict(
+            message="Unable to register elasticsearch percolate query {}.".format(uuid),
+            exception="Placeholder",
+            HTTPStatusCode=requests.codes.internal_server_error)),
+            requests.codes.internal_server_error)
+
+    return jsonify(dict(uuid=uuid)), requests.codes.created
 
 
 def delete(uuid: str, replica: str):
```
Comment (ttung) [inline dss/api/subscriptions.py]: `@dss_handler`

File: dss-api.yml (hunk 1 above)
Comment (Mackey22) [inline dss-api.yml]: This is part of the object returned by get

File: dss/api/subscriptions.py (hunk 2 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: 👍

File: dss/api/subscriptions.py (hunk 5 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: Yeah, it's a little funky. There may be a better way that I can look into

File: dss/api/subscriptions.py (hunk 3 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: 👍

File: dss/api/subscriptions.py (hunk 8 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: 👍

File: dss/api/subscriptions.py (hunk 7 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: Following format established in files.py: 

```
    return jsonify(
        dict(version=version)), requests.codes.created
```

File: dss/api/subscriptions.py (hunk 9 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: 👍

File: dss/api/subscriptions.py (hunk 11 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: Yeah that makes sense - I'll add that.

**Comment by Mackey22** [comment]
@ttung As a whole, I avoided rebasing against master because this branch is stacked on top of two other ones. Wanted to avoid a huge mess with the DSSExceptions and dss_handler decorator. Do you think I should anyways?

File: dss-api.yml (hunk 1 above)
Comment (kislyuk) [inline dss-api.yml]: Since nobody but the owner can see notifications for now, I question the need for including this field.

**Comment by ttung** [comment]
> @ttung As a whole, I avoided rebasing against master because this branch is stacked on top of two other ones. Wanted to avoid a huge mess with the DSSExceptions and dss_handler decorator. Do you think I should anyways?

Happy to show you how to rebase your entire stack if you want to.  It's a useful thing to be able to do.

Alternatively, you could do that rework on a separate PR.

File: dss/api/subscriptions.py [hunk 15]
```diff
@@ -3,35 +3,174 @@
...
+                           "Unable to register elasticsearch percolate query!")
+
+    extras['owner'] = owner
+    subscription_registration = _register_subscription(es_client, uuid, extras)
+    if subscription_registration['created']:
+        logger.debug(f"Event Subscription succeeded:\n{subscription_registration}")
+    else:
+        logger.critical(f"Event Subscription failed:\n{subscription_registration}")
+
+        # Delete percolate query to make sure queries and subscriptions are in sync.
```
Comment (mikebaumann) [inline dss/api/subscriptions.py]: To help ensure the percolate query gets removed if there is a problem adding the subscription, it seems try/except would be a useful construct here.
You were clearly careful with your error checking and I think this would pretty reliable as it stands, yet try/except is worth considering.

File: dss/api/subscriptions.py [hunk 16]
```diff
@@ -3,35 +3,174 @@
...
 def delete(uuid: str, replica: str):
-    pass
+    owner = request.token_info['email']
+
+    es_client = ElasticsearchClient.get(logger)
+    id_exists = es_client.exists(index=DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME,
+                                 doc_type=DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE,
+                                 id=uuid)
+    if not id_exists:
+        raise DSSException(requests.codes.not_found, "not_found", "Cannot find subscription!")
```
Comment (mikebaumann) [inline dss/api/subscriptions.py]: The nice pattern for doing this with one request to the ES service that you recently suggested to me could work well here, too.

File: dss/config.py [hunk 17]
```diff
@@ -61,13 +63,13 @@ def get_storage_schema(replica: str) -> str:
     @staticmethod
     def get_s3_bucket() -> str:
         if Config._S3_BUCKET is None:
-            if Config._CURRENT_CONFIG == BucketStage.NORMAL:
+            if Config._CURRENT_CONFIG == DeploymentStage.NORMAL:
```
Comment (mikebaumann) [inline dss/config.py]: I like the name change to DeploymentStage.

Comment (Mackey22) [inline dss/config.py]: Thanks!

File: dss/api/subscriptions.py (hunk 16 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: Yeah I'll convert to that

File: dss/api/subscriptions.py (hunk 15 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: Yeah that makes sense.

Are you sure there's an exception thrown if the subscription is not created? It seems logical that it would, but also weird that there would be a `created` field.

File: dss/api/subscriptions.py (hunk 5 above)
Comment (mikebaumann) [inline dss/api/subscriptions.py]: For Elasticsearch per se, and elasticsearch-py, '_source' is used very prominently and consistently in their docs. I am not aware of another option. The elasticsearch-dsl python library organizes its output differently and '_source' does not actually appear. The two output formats are different and each has its advantages and disadvantages.

File: chalice/app.py [hunk 18]
```diff
@@ -7,12 +7,12 @@
 pkg_root = os.path.abspath(os.path.join(os.path.dirname(__file__), 'chalicelib'))  # noqa
 sys.path.insert(0, pkg_root)  # noqa
 
-from dss import BucketStage, Config, create_app
+from dss import DeploymentStage, Config, create_app
```
Comment (ttung) [inline chalice/app.py]: I, too, like the rename, but this clutters the diff up significantly.  This should be done in a separate PR.  #290

File: dss/__init__.py [hunk 19]
```diff
@@ -77,16 +77,19 @@ def common_error_handler(exception):
         )
 
 class OperationWithAuthorizer(Operation):
-    authorized_domains = os.environ["AUTHORIZED_DOMAINS"].split()
+    Config.set_config(DeploymentStage.NORMAL)
```
Comment (ttung) [inline dss/__init__.py]: Why is this being set here?

File: dss/config.py [hunk 20]
```diff
@@ -97,20 +99,57 @@ def get_gs_bucket() -> str:
         return Config._GS_BUCKET
 
     @staticmethod
-    def _clear_cached_config():
+    def get_allowed_emails() -> str:
+        if Config._ALLOWED_EMAILS is None:
```
Comment (ttung) [inline dss/config.py]: Why is the allowed emails tied to the deployment stage?

File: dss/config.py [hunk 21]
```diff
@@ -97,20 +99,57 @@ def get_gs_bucket() -> str:
         return Config._GS_BUCKET
 
     @staticmethod
-    def _clear_cached_config():
+    def get_allowed_emails() -> str:
```
Comment (ttung) [inline dss/config.py]: should this be `get_allowed_domains()` or maybe `get allowed_email_domains()`?

File: dss/api/subscriptions.py (hunk 15 above)
Comment (mikebaumann) [inline dss/api/subscriptions.py]: I think an exception would not be thrown if a subscription is not created. It's more that the try block would cover for whatever else could go wrong it the subscription code that could exit the function without cleaning-up the query. It is straightforward and carefully written, so the chance of problems seems low, yet the try/except provides additional assurance and I think is the commonly expected construct when clean-up needs to be done if something goes wrong.

File: dss/api/subscriptions.py [hunk 22]
```diff
@@ -3,35 +3,174 @@
...
+from elasticsearch import Elasticsearch
+from elasticsearch.exceptions import NotFoundError
+from elasticsearch_dsl import Search
 from flask import jsonify, make_response, redirect, request
 from werkzeug.exceptions import BadRequest
 
-from .. import get_logger
-from .. import DSS_ELASTICSEARCH_INDEX_NAME, DSS_ELASTICSEARCH_DOC_TYPE, DSS_ELASTICSEARCH_QUERY_TYPE
-from .. import DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME, DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE
+from .. import (get_logger,
```
Comment (ttung) [inline dss/api/subscriptions.py]: If you want to make your code extra reviewable (particularly in the future), you can put a NL after your `(` and everything in subsequent lines will be indented +4, rather than be subject to the contents of the previous line.

Just a comment. Not a requirement.

File: dss/api/subscriptions.py [hunk 23]
```diff
@@ -3,35 +3,174 @@
...
 logger = get_logger()
 
 
+@dss_handler
 def get(uuid: str, replica: str):
-    pass
+    owner = request.token_info['email']
 
+    es_client = ElasticsearchClient.get(logger)
+    id_exists = es_client.exists(index=DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME,
```
Comment (ttung) [inline dss/api/subscriptions.py]: Is it required to do two round trips here?  Why not just do a get?

File: dss/api/subscriptions.py [hunk 24]
```diff
@@ -3,35 +3,174 @@
...
+                "properties": {
+                    "owner": {
+                        "type": "string",
+                        "index": "not_analyzed"
+                    }
+                }
+            }
+        }
+    }
+    get_elasticsearch_index(es_client, DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME, logger, index_mapping)
```
Comment (ttung) [inline dss/api/subscriptions.py]: I don't understand why you're getting an index mapping here. Help?

File: dss/api/subscriptions.py [hunk 25]
```diff
@@ -3,35 +3,174 @@
...
 def get(uuid: str, replica: str):
-    pass
+    owner = request.token_info['email']
 
+    es_client = ElasticsearchClient.get(logger)
+    id_exists = es_client.exists(index=DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME,
+                                 doc_type=DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE,
+                                 id=uuid)
+    if not id_exists:
+        raise DSSException(requests.codes.not_found, "not_found", "Cannot find subscription!")
```
Comment (ttung) [inline dss/api/subscriptions.py]: You should update your entry in the swagger spec to expect this possible return value.

File: dss/api/subscriptions.py [hunk 26]
```diff
@@ -3,35 +3,174 @@
...
+    response = es_client.get(index=DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME,
+                             doc_type=DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE,
+                             id=uuid)
+
+    source = response['_source']
+    source['uuid'] = uuid
+    source['replica'] = replica
+
+    if source['owner'] != owner:
+        raise DSSException(requests.codes.forbidden, "forbidden", "Your credentials can't access this subscription!")
```
Comment (ttung) [inline dss/api/subscriptions.py]: You should update your entry in the swagger spec to expect this possible return value.

File: dss/api/subscriptions.py [hunk 27]
```diff
@@ -3,35 +3,174 @@
...
+    }
+    get_elasticsearch_index(es_client, DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME, logger, index_mapping)
+
+    percolate_registration = _register_percolate(es_client, uuid, query)
+
+    if percolate_registration['created']:
+        logger.debug(f"Percolate query registration succeeded:\n{percolate_registration}")
+    else:
+        logger.critical(f"Percolate query registration failed:\n{percolate_registration}")
+        raise DSSException(requests.codes.internal_server_error,
```
Comment (ttung) [inline dss/api/subscriptions.py]: You should update your entry in the swagger spec to expect this possible return value.

File: dss/api/subscriptions.py (hunk 15 above)
Comment (ttung) [inline dss/api/subscriptions.py]: 👍

File: dss/api/subscriptions.py [hunk 28]
```diff
@@ -3,35 +3,169 @@
...
+                        "type": "string",
+                        "index": "not_analyzed"
+                    }
+                }
+            }
+        }
+    }
+    get_elasticsearch_index(es_client, DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME, logger, index_mapping)
+
+    try:
```
Comment (Mackey22) [inline dss/api/subscriptions.py]: @mikebaumann I'm going under the assumption that an exception is thrown if a document is not added. The example they give in the Elasticsearch documentation (https://www.elastic.co/guide/en/elasticsearch/guide/master/create-doc.html) of a document not being added responds with top-level objects of error and status. Status is 409, which translates nicely to Elasticsearch's Python Bindings (https://elasticsearch-py.readthedocs.io/en/master/exceptions.html) ConflictError. I think it's reasonable to assume Elasticsearch would throw this in this situation and another exception if for any other reason a doc can't be stored, so I'm catching all with ElasticsearchException.

File: tests/test_subscriptions.py [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+                                 id=uuid_)
+        registered_query = response['_source']
+        self.assertEqual(self.sample_percolate_query, registered_query)
+
+    def test_get(self):
+        find_uuid = self._put_subscription()
+        url = str(UrlBuilder()
+                  .set(path="/v1/subscriptions/" + str(find_uuid))
+                  .add_query("replica", "aws"))
+        with override_email_config(DeploymentStage.TEST):
```
Comment (ttung) [inline tests/test_subscriptions.py]: Why is it necessary to override email config to the test stage?  Shouldn't it be defaulting to the test stage in tests anyway?

File: tests/test_subscriptions.py [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+        uuid_ = self._put_subscription()
+
+        es_client = ElasticsearchClient.get(logger)
+        response = es_client.get(index=DSS_ELASTICSEARCH_INDEX_NAME,
+                                 doc_type=DSS_ELASTICSEARCH_QUERY_TYPE,
+                                 id=uuid_)
+        registered_query = response['_source']
+        self.assertEqual(self.sample_percolate_query, registered_query)
+
+    def test_get(self):
```
Comment (ttung) [inline tests/test_subscriptions.py]: should test your negative code paths like:

1. can't find it.
2. not the owner.

File: tests/test_subscriptions.py [hunk 31]
```diff
@@ -0,0 +1,157 @@
...
+        with override_email_config(DeploymentStage.TEST):
+            _, _, json_response = self.assertGetResponse(
+                url,
+                requests.codes.okay,
+                headers=self._get_auth_header())
+        self.assertEqual(self.sample_percolate_query, json_response['subscriptions'][0]['query'])
+        self.assertEqual(self.callback_url, json_response['subscriptions'][0]['callback_url'])
+        self.assertEqual(NUM_ADDITIONS, len(json_response['subscriptions']))
+
+    def test_delete(self):
```
Comment (ttung) [inline tests/test_subscriptions.py]: ditto re negative code paths.

File: chalice/app.py (hunk 18 above)
Comment (Mackey22) [inline chalice/app.py]: Sounds good. After you merge I'll rebase against that and that should take care of any conflicts

File: dss/__init__.py (hunk 19 above)
Comment (Mackey22) [inline dss/__init__.py]: Unnecessary. I'll remove it.

File: dss/api/subscriptions.py (hunk 22 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: 👍

File: dss/api/subscriptions.py (hunk 26 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: 👍

File: dss/api/subscriptions.py (hunk 24 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: I'll add a comment. Short explanation is that when I was testing locally, Elasticsearch was preprocessing inputs by splitting strings on punctuation. So for `john@example.com`, if I searched for people with the email address `jill@example.com`, `john@example.com` would show up because elasticsearch matched example w/ example. By including `"index": "not_analyzed"`, Elasticsearch leaves all owner inputs alone.

File: dss/config.py (hunk 21 above)
Comment (Mackey22) [inline dss/config.py]: Will change to `allowed_email_domains()`

File: dss/config.py (hunk 20 above)
Comment (Mackey22) [inline dss/config.py]: We're using Service Account credentials when testing on Travis but, for the moment, only allowing email addresses ending in `chanzuckerberg.com ucsc.edu broadinstitute.org ebi.ac.uk` when live.

File: tests/test_subscriptions.py (hunk 30 above)
Comment (Mackey22) [inline tests/test_subscriptions.py]: 👍

File: tests/test_subscriptions.py (hunk 29 above)
Comment (Mackey22) [inline tests/test_subscriptions.py]: Yep, this is unnecessary. I'll need to do this for negative code path 403 tho.

File: tests/test_subscriptions.py (hunk 31 above)
Comment (Mackey22) [inline tests/test_subscriptions.py]: 👍

File: dss/api/subscriptions.py (hunk 28 above)
Comment (mikebaumann) [inline dss/api/subscriptions.py]: That sounds right, @Mackey22. Thanks for reviewing this.

File: chalice/app.py (hunk 18 above)
Comment (ttung) [inline chalice/app.py]: I merged.  Please rebase!

Comment (Mackey22) [inline chalice/app.py]: :/

Comment (Mackey22) [inline chalice/app.py]: That actually wasn't nearly as bad as I expected it to be

Comment (ttung) [inline chalice/app.py]: Yeah, git's merge algorithms are pretty good.  It's good practice to do these no-op changes separately. :)

File: dss-api.yml [hunk 32]
```diff
@@ -345,6 +345,14 @@ paths:
                 pattern: "[A-Za-z0-9]{8}-[A-Za-z0-9]{4}-[A-Za-z0-9]{4}-[A-Za-z0-9]{4}-[A-Za-z0-9]{12}"
             required:
               - uuid
+        500:
+          description: Internal Server Error
+          schema:
+            $ref: '#/definitions/Error'
```
Comment (ttung) [inline dss-api.yml]: You should look at GET /files to see how you should pattern this.  It should include the allowed error codes.

Comment (Mackey22) [inline dss-api.yml]: K. Looks like I changed a couple of the paths and just forgot about this one.

Comment (Mackey22) [inline dss-api.yml]: Think I misunderstood. You want all of the error types to look like this:

```        default:
          description: Unexpected error
          schema:
            allOf:
              - $ref: '#/definitions/Error'
              - type: object
                properties:
                  code:
                    type: string
                    description: Machine-readable error code.  The types of return values should not be changed lightly.
                    enum: [unhandled_exception, not_found]
                required:
                  - code```
Where enum lists all of the possible error types? So there will just be one error listing for each endpoint?

File: dss/__init__.py [hunk 33]
```diff
@@ -9,6 +9,7 @@
 import json
 import logging
 
+import connexion
```
Comment (kislyuk) [inline dss/__init__.py]: I don't think this import is needed, please remove it.

File: dss/api/subscriptions.py [hunk 34]
```diff
@@ -3,35 +3,176 @@
...
+        response = es_client.get(index=DSS_ELASTICSEARCH_SUBSCRIPTION_INDEX_NAME,
+                                 doc_type=DSS_ELASTICSEARCH_SUBSCRIPTION_TYPE,
+                                 id=uuid)
+    except NotFoundError as ex:
+        raise DSSException(requests.codes.not_found, "not_found", "Cannot find subscription!")
+
+    source = response['_source']
+
+    if source['owner'] != owner:
```
Comment (kislyuk) [inline dss/api/subscriptions.py]: It's unclear in this notation which variable contains the owner in the request vs. the owner in the database document. Can you change it to be more explicit?

File: environment [hunk 35]
```diff
@@ -26,7 +26,10 @@ DSS_GS_BUCKET_TEST=org-humancellatlas-dss-test
 DSS_GS_BUCKET_TEST_FIXTURES=org-humancellatlas-dss-test-fixtures
 GOOGLE_APPLICATION_CREDENTIALS="${DSS_HOME}/gcp-credentials.json"
 API_HOST=hca-dss.czi.technology
-AUTHORIZED_DOMAINS="chanzuckerberg.com ucsc.edu broadinstitute.org ebi.ac.uk"
+TOKENINFO_URL=https://hca-dss.czi.technology/internal/tokeninfo
+DSS_SUBSCRIPTION_AUTHORIZED_DOMAINS="chanzuckerberg.com ucsc.edu broadinstitute.org ebi.ac.uk"
+DSS_SUBSCRIPTION_AUTHORIZED_DOMAINS_TEST="human-cell-atlas-travis-test.iam.gserviceaccount.com"
+DSS_SUBSCRIPTION_AUTHORIZED_DOMAINS_EMPTY=""
```
Comment (kislyuk) [inline environment]: As discussed offline - I don't think this variable or the NO_AUTH config stage is necessary

File: dss/api/subscriptions.py (hunk 34 above)
Comment (Mackey22) [inline dss/api/subscriptions.py]: 👍  Changed `owner` to `authenticated_user_email` and `source` to `stored_metadata`

File: dss/__init__.py (hunk 33 above)
Comment (Mackey22) [inline dss/__init__.py]: 👍

File: environment (hunk 35 above)
Comment (Mackey22) [inline environment]: Considering this and alternatives.

File: dss/config.py [hunk 36]
```diff
@@ -97,20 +99,57 @@ def get_gs_bucket() -> str:
...
+    try:
+        Config._CURRENT_CONFIG = temp_config
+        yield
+    finally:
+        Config._CURRENT_CONFIG = original_config
+        Config._clear_cached_bucket_config()
+
+
+@contextmanager
+def override_email_config(temp_config: DeploymentStage):
```
Comment (ttung) [inline dss/config.py]: This doesn't appear to be used / necessary any more.  Can this be nuked?

Alternatively, could this be used to fake the 403 instead of the flag?

Comment (Mackey22) [inline dss/config.py]: Yeah this can be nuked. We didn't want to use this to fake the 403 b/c a NO_AUTH deployment stage isn't the best. Gonna nuke this then merge
